### PR TITLE
mds: adding tunable features for caps_per_client

### PR DESCRIPTION
**Pull request by renhwztetecs**
Sometimes we need to raise or lower the value of "max_caps_per_client" and
"min_caps_per_client" to improve recall_client_state efficiency

Signed-off-by: huanwen ren <ren.huanwen@zte.com.cn>

### CODE REVIEW

File: src/mds/Server.cc [hunk 1]
```diff
@@ -1085,8 +1085,9 @@ void Server::recover_filelocks(CInode *in, bufferlist locks, int64_t client)
  */
 void Server::recall_client_state(float ratio)
 {
-  int max_caps_per_client = (int)(g_conf->mds_cache_size * .8);
-  int min_caps_per_client = 100;
+  int max_caps_per_client = (int)(g_conf->mds_cache_size * g_conf->max_ratio_caps_per_client);
+  int min_caps_per_client = (int)(g_conf->min_caps_per_client);
+  assert(min_caps_per_client < max_caps_per_client);
```
Comment (ukernel) [inline src/mds/Server.cc]: I think we should adjust improper numbers instead of crash

Comment (renhwztetecs) [inline src/mds/Server.cc]: Agree, I add the derr hint, and adjust max_caps_per_client.
thanks for you reviews.

File: src/mds/Server.cc [hunk 2]
```diff
@@ -1085,9 +1085,13 @@ void Server::recover_filelocks(CInode *in, bufferlist locks, int64_t client)
  */
 void Server::recall_client_state(float ratio)
 {
-  int max_caps_per_client = (int)(g_conf->mds_cache_size * .8);
-  int min_caps_per_client = 100;
-
+  int max_caps_per_client = (int)(g_conf->mds_cache_size * g_conf->max_ratio_caps_per_client);
+  int min_caps_per_client = (int)(g_conf->min_caps_per_client);
+  if (max_caps_per_client < min_caps_per_client) {
+    max_caps_per_client = min_caps_per_client + 1;
```
Comment (batrick) [inline src/mds/Server.cc]: This breaks a test: http://pulpito.ceph.com/pdonnell-2017-07-19_19:11:33-fs-wip-pdonnell-testing-20170718-distro-basic-smithi/1420077/

Key line is:

`2017-07-19T23:56:31.006 DEBUG:tasks.ceph_test_case:wait_until_equal: 202 != 80, waiting...
2017-07-19T23:56:33.637 INFO:tasks.ceph.mds.a.smithi031.stderr:2017-07-19 23:56:33.571051 7f4b5aab6700 -1 mds.0.server max_caps_per_client 101<min_caps_per_client 100`

The max_caps_per_client should be `80` which the test relies on. Please update the test to also change the min caps per client:

https://github.com/ceph/ceph/blob/359cd17ee98b73559d07c04a030a9506ab244000/qa/tasks/cephfs/test_client_limits.py#L45

File: src/mds/Server.cc [hunk 3]
```diff
@@ -1085,9 +1085,13 @@ void Server::recover_filelocks(CInode *in, bufferlist locks, int64_t client)
...
+    derr << "max_caps_per_client " << max_caps_per_client << "<" << "min_caps_per_client " << min_caps_per_client << dendl;
```
Comment (batrick) [inline src/mds/Server.cc]: Please change this to a normal debug message that prints **before** you change `max_caps_per_client`.

Comment (renhwztetecs) [inline src/mds/Server.cc]: updated it
thanks

File: src/mds/Server.cc (hunk 2 above)
Comment (renhwztetecs) [inline src/mds/Server.cc]: @batrick 
I modified it, please review it 
thank you  very much.

**Comment by renhwztetecs** [comment]
retest it please

**Comment by renhwztetecs** [comment]
jenkins retest this please

File: src/common/legacy_config_opts.h [hunk 4]
```diff
@@ -404,6 +404,8 @@ OPTION(client_metadata, OPT_STR, "")
 OPTION(client_acl_type, OPT_STR, "")
 OPTION(client_permissions, OPT_BOOL, true)
 OPTION(client_dirsize_rbytes, OPT_BOOL, true)
+OPTION(min_caps_per_client, OPT_INT, 100)
+OPTION(max_ratio_caps_per_client, OPT_FLOAT, .8)
```
Comment (batrick) [inline src/common/legacy_config_opts.h]: You can't add to this file anymore. Please add entries to `options.cc`.

File: src/mds/Server.cc [hunk 5]
```diff
@@ -1085,9 +1085,13 @@ void Server::recover_filelocks(CInode *in, bufferlist locks, int64_t client)
  */
 void Server::recall_client_state(float ratio)
 {
-  int max_caps_per_client = (int)(g_conf->mds_cache_size * .8);
-  int min_caps_per_client = 100;
-
+  int max_caps_per_client = (int)(g_conf->mds_cache_size * g_conf->max_ratio_caps_per_client);
```
Comment (batrick) [inline src/mds/Server.cc]: Now you access these values using: md_config_t::get().

Comment (renhwztetecs) [inline src/mds/Server.cc]: @batrick 
updated it, but not sure, so get the content is appropriate in the new config

File: src/common/legacy_config_opts.h (hunk 4 above)
Comment (renhwztetecs) [inline src/common/legacy_config_opts.h]: thanks for you hint

File: src/mds/Server.cc [hunk 6]
```diff
@@ -1085,8 +1086,14 @@ void Server::recover_filelocks(CInode *in, bufferlist locks, int64_t client)
  */
 void Server::recall_client_state(float ratio)
 {
-  int max_caps_per_client = (int)(g_conf->mds_cache_size * .8);
-  int min_caps_per_client = 100;
+  int max_caps_per_client = (int)(boost::lexical_cast<int>(g_conf->get_val<std::string>("mds_cache_size")) *
```
Comment (batrick) [inline src/mds/Server.cc]: use `g_conf->get_val<int>("mds_cache_size")`

Comment (renhwztetecs) [inline src/mds/Server.cc]: @batrick  thanks
changed it

**Comment by batrick** [comment]
Deferring this until after Luminous.

**Comment by renhwztetecs** [comment]
jenkins retest this please

**Comment by renhwztetecs** [comment]
jenkins retest this please

File: qa/tasks/cephfs/test_client_limits.py [hunk 7]
```diff
@@ -43,6 +43,11 @@ def _test_client_pin(self, use_subdir):
         open_files = 200
 
         self.set_conf('mds', 'mds cache size', cache_size)
+        min_caps_per_client = CAP_RECALL_MIN
+	max_caps_per_client = int(cache_size * CAP_RECALL_RATIO)
+	if max_caps_per_client < min_caps_per_client
```
Comment (batrick) [inline qa/tasks/cephfs/test_client_limits.py]: Mixed tabs/spaces.

File: src/common/options.cc [hunk 8]
```diff
@@ -5850,6 +5850,14 @@ std::vector<Option> get_mds_client_options() {
     .set_default(true)
     .set_description(""),
 
+    Option("min_caps_per_client", Option::TYPE_INT, Option::LEVEL_ADVANCED)
+    .set_default(100)
+    .set_description(""),
```
Comment (batrick) [inline src/common/options.cc]: New config options must have a description and docs added in `doc/cephfs/client-config-ref.rst`.

**Comment by renhwztetecs** [comment]
jenkins retest this please

File: src/common/options.cc (hunk 8 above)
Comment (renhwztetecs) [reply src/common/options.cc]: @batrick 
added it

File: qa/tasks/cephfs/test_client_limits.py (hunk 7 above)
Comment (renhwztetecs) [reply qa/tasks/cephfs/test_client_limits.py]: @batrick updated it
I think it's the editor's question, thank you

File: src/common/options.cc (hunk 8 above)
Comment (batrick) [reply src/common/options.cc]: @renhwztetecs The description in the above code needs updated.

Comment (rhwstatr) [reply src/common/options.cc]: @batrick updated it

Comment (renhwztetecs) [reply src/common/options.cc]: @batrick updated it

**Comment by batrick** [comment]
@renhwztetecs this needs a rebase.

**Comment by renhwztetecs** [comment]
jenkins retest this please

**Comment by renhwztetecs** [comment]
jenkins retest this please

**Comment by batrick** [comment]
@renhwztetecs Got a test failure:

````
2017-09-20T04:53:44.239 INFO:tasks.ceph.mds.b.smithi096.stderr:
2017-09-20T04:53:44.239 INFO:tasks.ceph.mds.b.smithi096.stderr:     0> 2017-09-20 04:53:43.828759 7fe9f0820700 -1 /build/ceph-13.0.0-1041-g46af6ca/src/common/config.h: In function 'T get_typed_value_visitor<T>::operator()(U&) [with U = double; typename boost::enable_if_c<(! boost::is_same<T1, T2>::value), int>::type <anonymous> = 0; T = float]' thread 7fe9f0820700 time 2017-09-20 04:53:43.824718
2017-09-20T04:53:44.240 INFO:tasks.ceph.mds.b.smithi096.stderr:/build/ceph-13.0.0-1041-g46af6ca/src/common/config.h: 331: FAILED assert("wrong type or option does not exist" == nullptr)
2017-09-20T04:53:44.240 INFO:tasks.ceph.mds.b.smithi096.stderr:
2017-09-20T04:53:44.240 INFO:tasks.ceph.mds.b.smithi096.stderr: ceph version 13.0.0-1041-g46af6ca (46af6caa761854e1022e213f41a4ebf30f83f15a) mimic (dev)
2017-09-20T04:53:44.240 INFO:tasks.ceph.mds.b.smithi096.stderr: 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x102) [0x55ebb9ab6b02]
2017-09-20T04:53:44.240 INFO:tasks.ceph.mds.b.smithi096.stderr: 2: (boost::detail::variant::invoke_visitor<get_typed_value_visitor<float> >::result_type boost::detail::variant::visitation_impl<mpl_::int_<0>, boost::detail::variant::visitation_impl_step<boost::mpl::l_iter<boost::mpl::l_item<mpl_::long_<8l>, boost::blank, boost::mpl::l_item<mpl_::long_<7l>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, boost::mpl::l_item<mpl_::long_<6l>, unsigned long, boost::mpl::l_item<mpl_::long_<5l>, long, boost::mpl::l_item<mpl_::long_<4l>, double, boost::mpl::l_item<mpl_::long_<3l>, bool, boost::mpl::l_item<mpl_::long_<2l>, entity_addr_t, boost::mpl::l_item<mpl_::long_<1l>, uuid_d, boost::mpl::l_end> > > > > > > > >, boost::mpl::l_iter<boost::mpl::l_end> >, boost::detail::variant::invoke_visitor<get_typed_value_visitor<float> >, void*, boost::variant<boost::blank, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned long, long, double, bool, entity_addr_t, uuid_d>::has_fallback_type_>(int, int, boost::detail::variant::invoke_visitor<get_typed_value_visitor<float> >&, void*, mpl_::bool_<false>, boost::variant<boost::blank, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned long, long, double, bool, entity_addr_t, uuid_d>::has_fallback_type_, mpl_::int_<0>*, boost::detail::variant::visitation_impl_step<boost::mpl::l_iter<boost::mpl::l_item<mpl_::long_<8l>, boost::blank, boost::mpl::l_item<mpl_::long_<7l>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, boost::mpl::l_item<mpl_::long_<6l>, unsigned long, boost::mpl::l_item<mpl_::long_<5l>, long, boost::mpl::l_item<mpl_::long_<4l>, double, boost::mpl::l_item<mpl_::long_<3l>, bool, boost::mpl::l_item<mpl_::long_<2l>, entity_addr_t, boost::mpl::l_item<mpl_::long_<1l>, uuid_d, boost::mpl::l_end> > > > > > > > >, boost::mpl::l_iter<boost::mpl::l_end> >*)+0x9c) [0x55ebb981892c]
2017-09-20T04:53:44.240 INFO:tasks.ceph.mds.b.smithi096.stderr: 3: (float md_config_t::get_val<float>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const+0x55) [0x55ebb9818a05]
2017-09-20T04:53:44.240 INFO:tasks.ceph.mds.b.smithi096.stderr: 4: (Server::recall_client_state()+0x5d) [0x55ebb97fc88d]
2017-09-20T04:53:44.241 INFO:tasks.ceph.mds.b.smithi096.stderr: 5: (MDCache::check_memory_usage()+0x41a) [0x55ebb9826aca]
2017-09-20T04:53:44.241 INFO:tasks.ceph.mds.b.smithi096.stderr: 6: (MDSRankDispatcher::tick()+0x215) [0x55ebb9772475]
2017-09-20T04:53:44.241 INFO:tasks.ceph.mds.b.smithi096.stderr: 7: (FunctionContext::finish(int)+0x2c) [0x55ebb976201c]
2017-09-20T04:53:44.241 INFO:tasks.ceph.mds.b.smithi096.stderr: 8: (Context::complete(int)+0x9) [0x55ebb975fbc9]
2017-09-20T04:53:44.241 INFO:tasks.ceph.mds.b.smithi096.stderr: 9: (SafeTimer::timer_thread()+0x45a) [0x55ebb9ab383a]
2017-09-20T04:53:44.241 INFO:tasks.ceph.mds.b.smithi096.stderr: 10: (SafeTimerThread::entry()+0xd) [0x55ebb9ab4c4d]
2017-09-20T04:53:44.241 INFO:tasks.ceph.mds.b.smithi096.stderr: 11: (()+0x770a) [0x7fe9f84dc70a]
2017-09-20T04:53:44.242 INFO:tasks.ceph.mds.b.smithi096.stderr: 12: (clone()+0x6d) [0x7fe9f754882d]
````

From: http://pulpito.ceph.com/pdonnell-2017-09-20_00:01:19-fs-wip-pdonnell-testing-20170919-testing-basic-smithi/1650332

The problem is the new options you added to `options.cc` are in the wrong option vector. You put them in the client options `get_mds_client_options` rather than under MDS `get_mds_options`.

**Comment by renhwztetecs** [comment]
@batrick 
thanks for reminding
I updated it

**Pull request merged**